numpy/_core/src/umath/stringdtype_ufuncs.cpp

```diff
@@ -0,0 +1,3063 @@
+
+        if (NpyString_pack(oallocator, ops, new_buf, strnlen(new_buf, s1.size)) < 0) {
+            npy_gil_error(PyExc_MemoryError, "Failed to pack string in strip");
+            goto fail;
+        }
+
+        PyMem_RawFree(new_buf);
+
+        in1 += strides[0];
+        in2 += strides[1];
+        out += strides[2];
+    }
+
+    NpyString_release_allocators(3, allocators);
+    return 0;
+
+fail:
+    NpyString_release_allocators(3, allocators);
+    return -1;
+
+}
+
+static NPY_CASTING
+strip_whitespace_resolve_descriptors(
+        struct PyArrayMethodObject_tag *NPY_UNUSED(method),
+        PyArray_DTypeMeta *NPY_UNUSED(dtypes[]),
+        PyArray_Descr *given_descrs[],
+        PyArray_Descr *loop_descrs[],
+        npy_intp *NPY_UNUSED(view_offset))
+{
+    Py_INCREF(given_descrs[0]);
+    loop_descrs[0] = given_descrs[0];
+
+    PyArray_Descr *out_descr = NULL;
+
+    if (given_descrs[1] == NULL) {
+        out_descr = (PyArray_Descr *)new_stringdtype_instance(
+                ((PyArray_StringDTypeObject *)given_descrs[0])->na_object,
+                ((PyArray_StringDTypeObject *)given_descrs[0])->coerce);
+
+        if (out_descr == NULL) {
+            return (NPY_CASTING)-1;
+        }
+    }
+    else {
+        Py_INCREF(given_descrs[1]);
+        out_descr = given_descrs[1];
+    }
+
+    loop_descrs[1] = out_descr;
+
+    return NPY_NO_CASTING;
+}
+
+static int
+string_lstrip_whitespace_strided_loop(
+        PyArrayMethod_Context *context,
+        char *const data[], npy_intp const dimensions[],
+        npy_intp const strides[], NpyAuxData *NPY_UNUSED(auxdata))
+{
+    PyArray_StringDTypeObject *descr = (PyArray_StringDTypeObject *)context->descriptors[0];
+    int has_string_na = descr->has_string_na;
+    int has_null = descr->na_object != NULL;
+    const npy_static_string *default_string = &descr->default_string;
+
+    npy_string_allocator *allocators[2] = {};
+    NpyString_acquire_allocators(2, context->descriptors, allocators);
+    npy_string_allocator *allocator = allocators[0];
+    npy_string_allocator *oallocator = allocators[1];
+
+    char *in = data[0];
+    char *out = data[1];
+
+    npy_intp N = dimensions[0];
+
+    while (N--) {
+        const npy_packed_static_string *ps = (npy_packed_static_string *)in;
+        npy_static_string s = {0, NULL};
+        int s_isnull = NpyString_load(allocator, ps, &s);
+
+        if (s_isnull == -1) {
+            npy_gil_error(PyExc_MemoryError, "Failed to load string in lstrip");
+            goto fail;
+        }
+
+        npy_packed_static_string *ops = (npy_packed_static_string *)out;
+
+        if (NPY_UNLIKELY(s_isnull)) {
+            if (has_string_na || !has_null) {
+                s = *default_string;
+            }
+            else {
+                npy_gil_error(PyExc_TypeError,
+                              "Cannot strip null values that are not strings");
+                goto fail;
```

Comment:
This seems odd?  A "proper" NA value (which I somewhat want to assume the user sets) should just propagate through elementwise functions, I would think?

Reply:
Good point, will audit the ufuncs for support for passing through NaN-like values.

Reply:
I almost finished this today but uncovered some other bugs, hopefully will have some fixups to push on Monday. Thanks for the push to expand the tests a bit!

Reply:
I finished this today. I ended up deciding to make ufuncs that output integers error for nan-like nulls. For ufuncs that return strings, I made them return nan-like nulls. Except for `replace`, which calls `count` internally and I had to add to the list of functions that error on nan-like nulls.

In practice hopefully people will be able to get around this by casting to e.g. `StringDType()` before calling one of these ufuncs.